### PR TITLE
[CORDA-2432] Update documentation for filtering by state relevancy in vault queries

### PR DESCRIPTION
**Pull request by JamesHR3**
In V4, the vault query criteria have been enhanced to allow filtering by state relevancy. The original default chosen was to return all states, as this is what happens in V3. However, this means that apps moving from targeting V3 to V4 that were just participants in the ledger require code changes in order to get the right behaviour. It's also unclear that this is the correct default for apps developed against V4.

This PR changes the default vault query criteria to return only relevant states for apps targeting V4. Apps targeting V3 will continue to see all states, as to do otherwise is a behaviour change.

Changes:
 - Make all variants of QueryCriteria with a relevancyStatus field have this default to RELEVANT.
 - For V3 apps, ensure that the query criteria used has relevancyStatus ALL.
 - Add a test that checks the default behaviour is correct if the CorDapp targets V3
 - Update the changelog to indicate that vault queries can be filtered by state relevancy.

### CODE REVIEW

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -320,6 +320,8 @@ Version 4.0
 
 * Logging for P2P and RPC has been separated, to make it easier to enable all P2P or RPC logging without hand-picking loggers for individual classes.
 
+* Vault Query Criteria have been enhanced to allow filtering by state relevancy. Queries can request all states, just relevant ones, or just non relevant ones. The default for apps targeting version 4 is to return only relevant states. Apps targeting version 3 will continue to see all states.
```
Comment (roger3cev) [inline docs/source/changelog.rst]: LGTM!

**Comment by JamesHR3** [comment]
@roger3cev I started adding some stuff to app-upgrade-notes.rst talking about how apps using observer node functionality will have to start using the relevancy criteria to get the right states, and realised that the vault query API docs also don't have anything about this. I'll update that too. Are you aware of anywhere else that should be updated? Is there anything other than the app upgrade nodes that talks about the differences developing for V4 as opposed to V3?

**Comment by rick-r3** [comment]
The end of the first paragraph in the description of this PR doesn't sound right to me: "However, this means that apps moving from targeting V3 to V4 that were just participants in the ledger require code changes in order to get the right behaviour. It's also unclear that this is the correct default for apps developed against V4."

Apps that assume queries return non-relevant (e.g. observers, apps doing unsupported stuff) will need to update their queries when upgrading their apps to targeting V4.

**Comment by JamesHR3** [comment]
You're right, I've not worded that very well - I was trying to describe the situation without the fix. (In other words, without doing this, apps that assumed only states relevant to them were returned would have to make changes when moving to target V4.)

To be clear then, with this fix in place:
 - Apps that assumed only relevant states were returned to them in vault queries (any app without observer functionality) will continue to receive the right states with no code changes
 - Apps that assumed non-relevant states were returned to them in vault queries (apps using observer node functionality) will need to update their vault queries in situations where they are interested in non-relevant states when moving to target V4.

**Comment by rick-r3** [comment]
Just thinking about this some more, maybe we can't do this.  An RPC client constructing queries doesn't know what the cordapp it's talking to is targeting, so a query from within a flow vs. via RPC may give different results and cause confusion (the criteria is built on the RPC client).

If the RPC client is V4, then the query result will be different to a V3 RPC client.  

Maybe changes like this just make the permutations to understand / test too complex?

**Comment by tudor-malene** [comment]
@rick-r3, just to add to your point.

An RPC client can have a different version of the cordapp on the classpath then the version that is installed on the node.  (Actually, I think on the RPC client you just need the contracts jar. )

Maybe what we want is a way to start the RPC client in "version 3 mode" or in "version 4 mode".

**Comment by JamesHR3** [comment]
Is the conclusion here that we can't move the default to RELEVANT, as it's too confusing to RPC clients? If so, I'll revert that change. We'll still need the documentation updates in here though (which I'll adjust accordingly).

**Comment by rick-r3** [comment]
Was just discussion with @tudor-malene some more.  I think it's best if we just remain consistent with V3 and leave this setting alone, and avoid lots of confusing paths that cause the behaviour to change.

File: docs/source/app-upgrade-notes.rst [hunk 2]
```diff
@@ -333,6 +333,16 @@ into shared business logic, but it makes perfect sense to put into a user-specif
 
 If your flows could benefit from being extended in this way, read ":doc:`flow-overriding`" to learn more.
 
+Step 10. Possibly update Vault state queries
+--------------------------------------------
+
+Queries made on a node's vault can filter by the relevancy of those states to the node in Corda 4. As this functionality does not exist in
+Corda 3, apps targeting that release will continue to receive all states in any vault queries. In Corda 4, the default is to return only those
+states relevant to the node when a vault query is made. For apps that only record states for which they are participants, no change is required
```
Comment (rick-r3) [inline docs/source/app-upgrade-notes.rst]: This section will need modifying to reflect to continuity of behaviour rather the change.  But it makes sense to encourage app authors to be more specific as other apps might start introducing non-relevant states.

File: docs/source/changelog.rst (hunk 1 above)
Comment (rick-r3) [reply docs/source/changelog.rst]: Needs updating to reflect we're keeping the default as is.

File: node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt [hunk 3]
```diff
@@ -831,6 +832,46 @@ abstract class VaultQueryTestsBase : VaultQueryParties {
...
+    @Test
+    fun `state relevancy queries`() {
+        database.transaction {
+            vaultFiller.fillWithSomeTestDeals(listOf("123", "456", "789"), includeMe = true)
+            vaultFiller.fillWithSomeTestDeals(listOf("ABC", "DEF", "GHI"), includeMe = false)
+            vaultFillerCashNotary.fillWithSomeTestCash(100.DOLLARS, notaryServices, 10, DUMMY_CASH_ISSUER, statesToRecord = StatesToRecord.ALL_VISIBLE)
+            vaultFillerCashNotary.fillWithSomeTestCash(100.DOLLARS, notaryServices, 10, DUMMY_CASH_ISSUER, charlie.party, statesToRecord = StatesToRecord.ALL_VISIBLE)
+            vaultFiller.fillWithSomeTestLinearStates(1, "XYZ", includeMe = true)
+            vaultFiller.fillWithSomeTestLinearStates(2, "JKL", includeMe = false)
+
```
Comment (rick-r3) [inline node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt]: Remove one of the blank lines

File: core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt [hunk 4]
```diff
@@ -265,14 +265,23 @@ sealed class QueryCriteria : GenericQueryCriteria<QueryCriteria, IQueryCriteriaP
...
+            owner: List<AbstractParty>? = null,
+            quantity: ColumnPredicate<Long>? = null,
+            issuer: List<AbstractParty>? = null,
+            issuerRef: List<OpaqueBytes>? = null,
+            status: Vault.StateStatus = Vault.StateStatus.UNCONSUMED,
+            contractStateTypes: Set<Class<out ContractState>>? = null
+        ) : this(
+                participants,
+                owner,
+                quantity,
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt]: Can this file just be reverted to how it looks on master?

**Pull request merged**